### ISSUE TITLE
Clang wrongly warns about empty variadic argument lists for macros

### ISSUE THREAD

**Issue by hasselmm** [llvm/llvm-project#76375]
Clang warns about empty variadic argument lists for macros if [-Wgnu-zero-variadic-macro-arguments](https://releases.llvm.org/17.0.1/tools/clang/docs/DiagnosticsReference.html#id380) is activated. This warning is wrong when compiling for C++ 20 or later mode. C++ 20 and later explicitly allow empty variadic argument lists for macros: https://eel.is/c++draft/cpp.replace.general#5

> If the [identifier-list](https://eel.is/c++draft/cpp.pre#nt:identifier-list) in the macro definition does not end with an ellipsis, the number of arguments (including those arguments consisting of no preprocessing tokens) in an invocation of a function-like macro shall equal the number of parameters in the macro definition[.](https://eel.is/c++draft/cpp.replace.general#5.sentence-1)
Otherwise, there shall be **at least as many arguments in the invocation as there are parameters in the macro definition (excluding the ...)**[.](https://eel.is/c++draft/cpp.replace.general#5.sentence-2)
There shall exist a ) preprocessing token that terminates the invocation[.](https://eel.is/c++draft/cpp.replace.general#5.sentence-3)

**Comment by hazohelet** [comment]
Could you provide a reproducer?

Clang code seems to avoid emitting `Wgnu-zero-variadic-...` if you specify C++20 or later.
https://github.com/llvm/llvm-project/blob/1022febd9df30abbd5c490b94290c4422ca15b01/clang/lib/Lex/PPMacroExpansion.cpp#L996-L1002

**Comment by hasselmm** [comment]
The following snippet:
``` C++
#define TEST(Type, Name, ...) auto Name = Type{#Name, ##__VA_ARGS__};
struct Test { Test(const char *); };
TEST(Test, test);
```
fails like this:
```
$ clang++ -c test.cpp -std=c++20 -pedantic
test.cpp:1:53: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
    1 | #define TEST(Type, Name, ...) auto Name = Type{#Name, ##__VA_ARGS__};
      |                                                     ^
1 warning generated.
```

**Comment by shafik** [comment]
So the diagnostic is `ext_paste_comma` and it looks like the issue is the comma before not that it is empty. See

https://github.com/llvm/llvm-project/blob/98073057ee45bc4a9256ecc70dc3da20b0fd5d3b/clang/lib/Lex/TokenLexer.cpp#L152-L167

and

https://github.com/llvm/llvm-project/blob/98073057ee45bc4a9256ecc70dc3da20b0fd5d3b/clang/lib/Lex/TokenLexer.cpp#L515-L525

CC @cor3ntin @tahonermann

**Comment by seanm** [comment]
I'm seeing this too. Seems to still happen in current master.
